(PUP-5648) Add Iterable and Iterator to Puppet Type System

thallgren opened the pull request:
This PR adds the new types `Iterable` and `Iterator` to the Puppet
type system together with the runtime support needed to represent an
object that has been transformed into an Iterator by a call to
`Iterable.on`. This method transforms an instance of an object that
Puppet can use in iterative functions into an Iterator. The Iterator
in turn, provides basic support for the iterative functions.

The PR also changes the iterative functions `each`, `filter`,
`map`, `reduce`, and `slice` so that the type of their declared
_enumerable_ parameter is `Iterable` instead of `Any`

hlindberg commented:
This must be broken up into multiple commits:

- adding the iterable type with related tests
- one commit per existing function switching to support iterator
- one commit per added function

I can also see breaking this up into multiple tickets; the features do need separate documentation etc.

hlindberg commented on lib/puppet/functions/reverse_each.rb:
```diff
@@ -0,0 +1,88 @@
+# Reverses the order of the elements of something that is iterable and optionally runs a
```
We must check if it is ok to just define the 4.x function, or if we have to do a 3.x version only to get the documentation generated - ping @nfagerlund

hlindberg commented on lib/puppet/functions/reverse_each.rb:
```diff
@@ -0,0 +1,88 @@
+# # named after the item's value and uses the item's value in a parameter.
+# ~~~
+#
+# When no second argument is present, Puppet returns a new iterable so that a new function that takes
```
When no block is given...

hlindberg commented on lib/puppet/functions/reverse_each.rb:
```diff
@@ -0,0 +1,88 @@
+#
+Puppet::Functions.create_function(:reverse_each) do
+  dispatch :reverse_each do
+    param 'Any', :iterable
```
Can we have tighter checking here instead of delegating to Types::Iterable? Specific types would also show up in documentation (which is better than seeing Any).

thallgren replied:
I like the concept of being able to describe that something is transformable into something else. Could you create a specific PR for that? And meanwhile, perhaps we could keep Any here since it's rather complicated to describe with a Variant.
```
Variant[Integer,Type[Integer[<with range spec>],String[1],Array[1],Hash[1],Iterable,...]
```
Perhaps we'd like to make the Type[Tuple] and a and Type[Struct] Iterable? If we do, then that too (and any other changes we make in the future), will have an impact on the documentation of this method.

Also, for the sake of consistency, `Any` is used as the type for the enumerable in the functions `each`, `filter`, `reduce`, and `map` so a tighter checking here should imply a tighter checking in  them too.

hlindberg commented on lib/puppet/functions/reverse_each.rb:
```diff
@@ -0,0 +1,88 @@
+  end
+
+  dispatch :reverse_each_block do
+    param 'Any', :iterable
```
Same here

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+#
+Puppet::Functions.create_function(:step) do
+  dispatch :step do
+    param 'Any', :iterable
```
same here; Any

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+#
+# @example Using the `step` function
+#
+# `$data.step(<n>) |$parameter| { <PUPPET CODE BLOCK> }`
```
needs to be concrete since it is an example - no <n> or <PUPPET CODE BLOCK>.

thallgren replied:
Using "or" in examples seems to have been widely accepted. I followed the style that was used in `each`, `filter`,  `map`, and `reduce` to make the documentation of these functions consistent and easy to compare.

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+#
+# `$stepped_data = step($data, <n>)`
+#
+# When no third argument is present, Puppet returns an iterable that represents the nth element
```
When no block is given

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# `$stepped_data = step($data, <n>)`
+#
+# When no third argument is present, Puppet returns an iterable that represents the nth element
+# of its first argument. This allows methods on iterables to be chained.
```
change "methods" to "functions"

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# When no third argument is present, Puppet returns an iterable that represents the nth element
+# of its first argument. This allows methods on iterables to be chained.
+#
+# When a lamdba is given as the third argument, Puppet iterates the first argument and passes each
```
lambdas are not given as arguments in the perspective of a user - it is a given block. Rephrase.

thallgren replied:
I agree that it should be called a block and that is't not an argument. In general though, I find a lot of the function documentations confusing because many of the mentions that the function takes <n> "arguments" and one of the arguments is often a "lambda" (see `each` for instance). Perhaps we should have a JIRA on how to be more consistent about this throughout in the documentation.

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# of its first argument. This allows methods on iterables to be chained.
+#
+# When a lamdba is given as the third argument, Puppet iterates the first argument and passes each
+# nth value in turn to the lambda, then returns the first argument unchanged.
```
use "block" instead of lambda

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# When a lamdba is given as the third argument, Puppet iterates the first argument and passes each
+# nth value in turn to the lambda, then returns the first argument unchanged.
+#
+# @example Using the `step` function with an array, a step factor, and a one-parameter lambda
```
here lambda is fine, since it is on the calling side, the mechanism with which a block is given to the function. (Pointing this out just to avoid the otherwise obvious question).

thallgren replied:
My preference would be to use the term block throughout. The distinction between the terms block and lambda is fairly subtle and using both doesn't make things clearer. Actually, I think there's a risk that it adds confusion.

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# @example Using the `step` function with an array, a step factor, and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the array $data, run a lambda that creates a resource in reverse order for each item in reverse.
```
"run a lambda" => "call a lambda". The description seems wrong - the lambda does not reverse (copy/paste error?), the lambda just notices. Here it is the step function that picks every 3:d element that the lambda notices.

thallgren replied:
Oops. Definitely a copy paste error.

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# $data.step(3) |$item| {
+#  notice($item)
+# }
+# # Puppet will output 3 notices with the texts '1', '4', '7'.
```
The typical documentation language for this would be "Puppet notices the values '1', '4' and '7'" (i.e. the concept of "noticing" is known to the reader).

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+# # Puppet will output 3 notices with the texts '1', '4', '7'.
+# ~~~
+#
+# When no lambda is present, Puppet returns a new iterable so that a new function that takes
```
"When no lambda is given", or "When no block is present" (depending on point of view).

thallgren replied:
Uhm, in your comment above you wanted "When no block is given" (which sort of proves my point on using both terms). I think I'll stick with that.

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+#
+# @example Using the `step` function chained with a `map` function.
+#
+# # For the array $data, return an array containing each value multiplied by 10 in reverse order
```
probably better to say "each value in reverse order multiplied by 10" (as you may otherwise wonder how a "multiplication in reverse" is performed.

thallgren replied:
Yes, or even better, "each 5th value in reverse order multiplied by 10" :-)

hlindberg commented on lib/puppet/functions/step.rb:
```diff
@@ -0,0 +1,88 @@
+#
+# @example The same example using `step` function chained with a `map` in alternative syntax
+#
+# # For the array $data, return an array containing each value multiplied by 10 in reverse order
```
same remark as on the previous example.

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,70 @@
-      x.is_a?(Array) ? x.each : x
-    when Integer
-      o.times
+      EnumerationIterable.new(PStringType::DEFAULT, o.each_char)
```
Does this work on 1.9.3? (I suppose travis will tell)

hlindberg replied:
Yes it works. Test is green on 1.9.3

hlindberg commented on spec/shared_behaviours/iterative_functions.rb:
```diff
@@ -67,3 +59,25 @@
+  end
+end
+
+shared_examples_for 'all iterative functions that requires no block' do |func|
```
does this mean they optionally accept a block, or that they have no opinion what so ever about a block?

thallgren replied:
In this context it means that they do not require a block. Whether or not they optionally accepts one isn't tested here.

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+
+require 'shared_behaviours/iterative_functions'
+
+describe 'the reverse_each method' do
```
"method" => "function"

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+describe 'the reverse_each method' do
+  include PuppetSpec::Compiler
+
+  it 'raises an error when used against an unsupported type' do
```
"raises an error when given an unsupported type"

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    end.to raise_error(Puppet::Error, /Cannot be coerced into an Iterable/)
+  end
+
+  it 'raises an error when called with any parameters besides a block' do
```
"parameters" => "arguments"

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+      compile_to_catalog(<<-MANIFEST)
+        [1].reverse_each(1)
+      MANIFEST
+    end.to raise_error(Puppet::Error, /expects 1 argument, got 2/)
```
This test looks strage - it fails on getting two arguments (ike example above), not on being given something that is not a block) - you cannot really do that in puppet source as the only way to get a block there is to specify one.

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    expect(eval_and_collect_notices('Integer[5,8].reverse_each |$x| { notice($x) }')).to eq(['8', '7', '6', '5'])
+  end
+
+  it 'should produce negative nth element of range Integer[5,20] when chained after a step' do
```
decending ? (n is never negative)

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    expect(eval_and_collect_notices('Integer[5,20].step(4).reverse_each |$x| { notice($x) }')).to eq(['17', '13', '9', '5'])
+  end
+
+  it 'should produce negative nth element of times 5 when chained after a step' do
```
descending

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    expect(eval_and_collect_notices('5.step(2).reverse_each |$x| { notice($x) }')).to eq(['4', '2', '0'])
+  end
+
+  it 'should produce negative nth element of range Integer[5,20] when chained after a step' do
```
descending

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  it 'should produce negative nth element of range Integer[5,20] when chained after a step' do
+    expect(eval_and_collect_notices('[5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20].step(4).reverse_each |$x| { notice($x) }')).to eq(['17', '13', '9', '5'])
```
long line, please break it up

hlindberg commented on spec/unit/functions/reverse_each_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    expect(eval_and_collect_notices('[5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20].step(4).reverse_each |$x| { notice($x) }')).to eq(['17', '13', '9', '5'])
+  end
+
+  it_should_behave_like 'all iterative functions that requires no block', 'reverse_each'
```
suggested name change since "requires no block" is not very clear

hlindberg commented on spec/unit/functions/step_spec.rb:
```diff
@@ -0,0 +1,93 @@
+describe 'the step method' do
+  include PuppetSpec::Compiler
+
+  it 'raises an error when used against an unsupported type' do
```
"raises an error when give an argument of unsupported type" (or even better), "of a type that cannot be iterated"

hlindberg commented on spec/unit/functions/step_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end.to raise_error(Puppet::Error, /Cannot be coerced into an Iterable/)
+  end
+
+  it 'raises an error when called with any parameters besides a block' do
```
"parameters" => "arguments"

hlindberg commented on spec/unit/functions/step_spec.rb:
```diff
@@ -0,0 +1,93 @@
+      compile_to_catalog(<<-MANIFEST)
+        [1].step(1,2)
+      MANIFEST
+    end.to raise_error(Puppet::Error, /expects 2 arguments, got 3/)
```
as before this test is questionable - unless tested with a call that is constructed in ruby

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,121 @@
+module Puppet::Pops::Types
+describe 'The enumeration support' do
+  it 'produces an iterable for Hash' do
+    expect(Iterable.on({:a=>1})).to respond_to(:step)
```
Can this be made more precise? Or is the entire Iterable API based on object having a step method?

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,121 @@
+    expect{ |b| Iterable.on(2..7).each(&b) }.to yield_successive_args(2,3,4,5,6,7)
+  end
+
+  it 'produces correct element type for an iterable on a integer range' do
```
replace "correct" with what it explicitly does

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,121 @@
+    expect{ |b| Iterable.on(PIntegerType.new(2,7)).each(&b) }.to yield_successive_args(2,3,4,5,6,7)
+  end
+
+  it 'produces correct element type for an iterable on bounded Integer type' do
```
"correct" is vague, be explicit

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,121 @@
+    expect{ |b| Iterable.on('eat this').each(&b) }.to yield_successive_args('e', 'a', 't', ' ', 't', 'h', 'i', 's')
+  end
+
+  it 'produces correct element type for an iterable on Array' do
```
"correct" is vague

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,121 @@
+    expect{ |b| Iterable.on([1,5,9]).each(&b) }.to yield_successive_args(1,5,9)
+  end
+
+  it 'produces correct element type for an iterable on Array' do
```
"correct"

hlindberg commented on spec/unit/pops/types/type_calculator_spec.rb:
```diff
@@ -234,6 +235,14 @@ def type_from_class(c)
+    it 'iterable translates to PIterableType' do
+      expect(calculator.infer(Puppet::Pops::Types::Iterable.on(1))).to be_a(Puppet::Pops::Types::PIterableType)
+    end
+
```
As noted earlier, it seems better to have two types; `Iterable` - which includes all the kinds of things for which an Iterator can be created, and `Iterator`, the thing that iterates. That way, the functions can use  a signature of `Iterable` instead of `Any`, and we can document that `Iterable` matches String, Integer, Array, Hash, and Iterator (and later possibly Range which we do not yet have as a runtime type in the Puppet Language, nor in the type system).

thallgren commented:
PR updated so that:

1. It now only adds the `Iterable` type. No new functions are added.
2. It's divided into two commits. One that adds the type and one that removes the many times duplicated method `asserted_enumerable` in favor of one single method declared on the `Enumeration` class.

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
   def self.enumerator(o)
     @@singleton ||= new
     @@singleton.enumerator(o)
   end
```
I wonder why I did it like this. Don't think we need a singleton as opposed to just using the class method `enumerator` directly. Can't see we want to use different implementations at runtime, so no need to pass one around etc.

thallgren replied:
I also reacted to this but decided not to change it since it's not related to this PR.

hlindberg replied:
ok, we can clean that up later.

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
-    when Puppet::Pops::Types::PIntegerType
-      # Not enumerable if representing an infinite range
-      o.enumerable? ? o.each : nil
+      EnumerationIterable.new(TypeCalculator.singleton.infer_and_reduce_type(o), o.each)
```
Is this really what is wanted? (infer_and_reduce). Consider the resulting type of `[1,2, 101, 102]`. I think this will be `Iterable[Integer[1,102]`. Which will then not match the more precisely specified `Interable[Variant[Integer[1,10]], Variant[Integer[100,110]]`. If the Iterable retained the full inference it would be of more value.

thallgren replied:
This is how the TypeCalculator will infer the element type of an Array, and hence, also how the element type of the Iterator is calculated. It could be improved of course (the infer_and_reduce method), but it's not in scope for this PR IMO.

hlindberg replied:
But there is another way of doing the infer that retains the information of all the values. That is only possible when inferring from values. i.e. why not use `infer_set` and skip reduction to common single element type?

thallgren replied:
The motivation for the current implementation is consistency. An `Array[T]` should yield `Iterable[T]` in my opinion. Do you see that differently?

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
-      o.enumerable? ? o.each : nil
+      EnumerationIterable.new(TypeCalculator.singleton.infer_and_reduce_type(o), o.each)
+    when Integer
+      o > 0 ? IntegerRangeIterator.new(PIntegerType.new(0, o - 1)) : nil
```
I think this means that an integer behaves as if you did `n.times`. Not sure that is great since if an integer contains something to iterate over, then it would be its bits. Seeing something like this:
```
3.each [ $x | { notice $x }
```
is confusing - is that an iteration of only [3], or [1,2,3], or [0,1,2,3], or [0,1,2]? (I see the impl is the range 0..(n-1).
Think a `times´ function is needed instead of directly getting an enumerator.

thallgren replied:
The code used to read:
```
    when Integer
      o.times
```
I totally agree that a dedicated `times` function would much better. This PR doesn't change this behavior however. A separate JIRA perhaps?

hlindberg replied:
Ah, I misread since these lines looked like they were added, So I thought it was a new feature.
Since it is there already... :hear_no_evil: :see_no_evil: :speak_no_evil:

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def enumerator(o)
     case o
```
It would be nice to add iteration of all PTypes that contain types (but that is a separate ticket).

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
+      else
+        # A generalized form of the inferred type for the 'min' is used here since inferring the full
+        # range might waste a lot of memory.
+        EnumerationIterable.new(TypeCalculator.singleton.infer(min).generalize, o.each)
```
if type is a string, you know that the length of the resulting strings will be between length of the shortest, and length of the longest. To support this in the puppet language we are going to need a PRange type as well, but nice to infer a slightly richer type for strings here. (It will be used since people do generate sequences of unique names).

For Float, the min, max are known. 

Ruby allows creation of ranges like this `([1]..[10])`, which is kind of useless except as a Tuple of two values since most default range operations do not work - it will fail on the `o.each` (that is probably good enough protection against bad stuff).

hlindberg replied:
@thallgren easy to add better inference here for Range of String, Integer, Float.  I think it is easier to add support for those here now, that to do a separate ticket and all the admin around that.

I played with enumeration of Ruby Range, it it errors on [1..3.13], or ["a"..10] etc. The infer and reduce here is therefore not that valuable.

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
     else
-      nil
+      begin
+        result = o.each
```
responds_to? is better here, since you may otherwise hide an error inside of the each implementation if it calls an undefined method.

hlindberg replied:
@thallgren ping

hlindberg commented on lib/puppet/pops/types/enumeration.rb:
```diff
@@ -1,33 +1,72 @@
-      nil
+      begin
+        result = o.each
+        result.is_a?(Enumerator) ? result : nil
```
Hm, we have no type information here, requires you to accept `Iterable[Any]`. Do we really need this very open ended solution? Can we require a richer API? As an example, if we allow producer functions, and we have return types, we do know the produced type. Perhaps have either another method enumerator(o, T) that you need to use if what you created does not have type information. (That implies that the values that will be emitted needs to be type checked before yielded).

thallgren replied:
Let's discuss. I added this as a fallback. Thought it would be better than to just bail out. Perhaps it isn't.

hlindberg replied:
I think it is better to bail out since there is loss of fidelity in that last case anyway, and that can cause problems if there are other mistakes made elsewhere that makes it end up here.

thallgren commented:
Pushed a new commit that I believe addresses all concerns, both regarding too relaxed element types and uncertainties on what types that are supported. That set is now very distinct.

hlindberg commented on lib/puppet/pops/types/iterable.rb:
```diff
@@ -0,0 +1,291 @@
+    #
+    # @param o [Object] The object to produce an `Iterable` for
+    # @return [Iterable,nil] The produced `Iterable` or `nil` if it couldn't be produced
+    # @raise [ArgumentError] In case an `Iterable` cannot be produced
```
This contradicts the line above

hlindberg commented on lib/puppet/pops/types/iterable.rb:
```diff
@@ -0,0 +1,291 @@
+          # Unsupported range. It's either descending or nonsensical for other reasons (float, mixed types, etc.)
+          nil
+        end
+      when Dir
```
Are you using this for something? Otherwise I see no reason to add this.

thallgren replied:
It was the only implementation in Ruby that we doesn't otherwise cover that seemed to make sense. I can remove it.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -90,6 +94,29 @@ def kind_of_callable?(optional=true)
+        false
+      end
+
+      # Returns the `PIterableType` that this type should be assignable to to, or `nil` if no such type exists.
```
typo "to to"

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -90,6 +94,29 @@ def kind_of_callable?(optional=true)
+      # A String is assignable to an Iterable[String] iterating over the strings characters
+      # An Integer is assignable to an Iterable[Integer] iterating over the 'times' enumerator
+      # A Type[T] is assignabel to an Iterable[Type[T]] if T is an Integer or Enum
+      #
```
These are just examples, since the actual implementation that makes these true are elsewhere (in the respective types). Maybe point that out.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -203,6 +230,29 @@ def hash
+        when PEnumType
+          true
+        when PIntegerType
+          @type.enumerable?
```
isn't that method deprecated?  - Should this be `@type.iterable?`

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -203,6 +230,29 @@ def hash
+        when PEnumType
+          @type.each
+        when PIntegerType
+          @type.enumerable? ? @type.each : nil
```
same here, iterable? vs. enumerable?

thallgren replied:
No. `iterable?` and `enumerable?` has different meaning (type versus instances of the type). I did add specific method documentation to explain what they do.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -378,12 +426,17 @@ def initialize(values)
+      end
+
+      def iterable_type
+        PIterableType.new(PStringType::DEFAULT)
```
Doesn't this lose precision? For an Enum, the strings are known. so it is possible to do either Iterable[String[min,max]], or why not just Iterable[<this enum>] ?

thallgren replied:
It should be changed to `Iterable[String[1,1]]` since an iterable on a string yields all it's characters.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+          when Iterable, String, Hash, Enumerable, Enumerator, Range, PEnumType
+            true
+          when Integer
+            o > 0
```
Shouldn't this be `>= 0` ? 0.times is just as valid as iterating over an empty array

thallgren replied:
Agree. This is inconsistent. Will change.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Any element_type will do
+          case o
+          when Iterable, String, Hash, Enumerable, Enumerator, Range, PEnumType
```
why are Enumerable and Enumerator here? Shouldn't this be Array? We are not really supporting any Ruby type that can be enumerated since we do not know what they return. Although we have the same problem with Hash and Array we have chosen to use the Ruby objects as Puppet runtime objects, so we have to live with those.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+          when Integer
+            o > 0
+          when PIntegerType
+            o.enumerable?
```
Why enumerable and not iterable? Is this asking about the Ruby quality enumerable as opposed to the Puppet qulity of being iterable? (Confusing)

thallgren replied:
See comment above regarding enumerable? versus iterable?

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Don't request the iterable_type since it might be expensive to compute the
+          # type of its elements
+          o.iterable?
```
Hm, don't understand how it can accept any iterable

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+      def _assignable?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Don't request the iterable_type since it might be expensive to compute the
+          # type of its elements
```
I got tricked here. Worth adding this to the comment ", since this Iterable accepts Any element, it is enough that o is iterable"

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+      end
+
+      def instance?(o)
+        o.is_a?(Iterable) && (@element_type.nil? || @element_type.assignable?(o.element_type))
```
How do you know that the `o` in `o.element_type` has such a member?

thallgren replied:
Because `o.is_a?(Iterable)` and `Iterable` defines the method `element_type`.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
+
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PIteratorType) && (@element_type.nil? || @element_type.assignable?(o.element_type))
```
same thing here with `o.element_type`

thallgren replied:
Because `o.is_a?(PIteratorType)` and `PIteratorType` defines the method `element_type`

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -621,6 +794,14 @@ def hash
+      end
+
+      def iterable_type
+        PIterableType.new(PStringType::DEFAULT)
```
I don't get this, why are all these not using the information in the type?

thallgren replied:
Yes, that would be an improvement. Will fix.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -840,11 +1021,26 @@ def generalize
+
+      def iterable_type
+        if self == DEFAULT
+          PIterableType.new(PTupleType.new([PAnyType::DEFAULT], PHashType::TUPLE_SIZE))
```
Do not understand, does this create an Iterable[Iterable[Any, what?? ) Think a comment is needed here

hlindberg replied:
Is this for the type `Struct[{}]` - If having one of those it does not really matter what the elementy type is, anything goes. I should use Tuple[Unit, Unit] I think (i.e. "whatever you want the type to be").

hlindberg replied:
Reading again, I think this is for any hash, and then it looks ok.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -840,11 +1021,26 @@ def generalize
+          PIterableType.new(PTupleType.new([PAnyType::DEFAULT], PHashType::TUPLE_SIZE))
+        else
+          tc = TypeCalculator.singleton
+          key_type = tc.infer_and_reduce_type(@elements.map {|se| se.key_type })
```
why reduce and generalize? doesn't that lose information

thallgren replied:
Oversight. This should be changed to use the same improved logic that was added to Iterable.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -840,11 +1021,26 @@ def generalize
+        else
+          tc = TypeCalculator.singleton
+          key_type = tc.infer_and_reduce_type(@elements.map {|se| se.key_type })
+          value_type = tc.infer_and_reduce_type(@elements.map {|se| se.value_type })
```
same here

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -983,6 +1179,14 @@ def instance?(o)
+      end
+
+      def iterable_type
+        PIterableType.new(TypeCalculator.singleton.infer_and_reduce_type(types))
```
lossy, this is really Iterable[Variant [each element_type in tuple]]

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -1263,6 +1467,18 @@ def instance?(o)
+        if self == DEFAULT || self == EMPTY
+          PIterableType.new(PTupleType.new([PAnyType::DEFAULT], TUPLE_SIZE))
+        else
+          PIterableType.new(PTupleType.new([@key_type, element_type], TUPLE_SIZE))
```
is element_type a typo? or is it element_type() (a call), or should it be @element_type, etc...

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -1272,6 +1488,7 @@ def is_the_empty_hash?
       end
 
       DEFAULT = PHashType.new(nil, nil)
+      TUPLE_SIZE = PIntegerType.new(2,2)
```
Don't get the naming here. What TUPLE_SIZE?

thallgren replied:
The size of a TUPLE in a key pair. What other tuples are there :-). Seriously. I'll change it to KEY_PAIR_TUPLE_SIZE.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -1307,8 +1524,9 @@ class PVariantType < PAnyType
+      # @param types [Array[PAnyType]] the variants
       def initialize(types)
-        @types = types.freeze
+        @types = types.uniq.freeze
```
Other optimizations here would be to reduce the set based on assignability. e.g:
`Variant[Integer, Float, Any]`, that is just Variant[Any], or infact, Any :-)

thallgren replied:
Assume you do infer an element type for [0,1,1,1,1,1,1,2,2,2,2,2,2]. Do you really want that to result in a Variant with 13 Integer types? Or is 3 enough?

hlindberg replied:
ok, it is fine to have it here, but it should really be in the others since it seems it would waste cycles first creating types for all of the integers, and then they are made unique. In your example case it could even be optimized into Integer[0,2]. We should perhaps use a Set when creating the set of types for Variant (not sure how wan avoid creating the type to add though, so we still end up with garbage). --- i.e. leave it like you have it in this PR. May want to optimize some of the inferences later.

hlindberg commented on lib/puppet/pops/types/types.rb:
```diff
@@ -1394,6 +1612,15 @@ def instance?(o)
+      def iterable?
+        c = class_from_string(@runtime_type_name)
+        c.nil? ? false : c < Enumerable || c < Iterable
+      end
```
think this makes any Ruby Enumerable be Iterable. Don't think we want that. It is too open ended.

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,229 @@
+module Puppet::Pops::Types
+describe 'The iterable support' do
+
+  [5, (3..10), %w(a b c), 'hello', PIntegerType.new(1, 4), PEnumType.new(%w(yes no)) ].each do |obj|
```
suggest formatting each input case on a separate row - makes it easier to spot the last line
`].each do |obj| ` and to identify each case

hlindberg replied:
shouldn't there be a hash there?

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,229 @@
+    end
+  end
+
+  { -1 => 'a negative Integer', 0 => 'the value 0', PIntegerType.new(nil, nil) => 'an unbounded Integer type' }.each_pair do |obj, desc|
```
one line per input case

hlindberg replied:
This list needs a Ruby class that is Eumerable. To test that we do not accept just any Enumerable.

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,229 @@
+
+  it 'produces an iterable with element type String[1,1] for an iterable created on a String' do
+    expect(Iterable.on('eat this').element_type).to eq(PStringType.new(PIntegerType.new(1,1)))
+ end
```
so I was wrong earlier this example is correct; which means I misunderstood what the code was for...

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,229 @@
+    expect(Iterable.on([1,5,5,9,9,9]).element_type).to eq(PVariantType.new([PIntegerType.new(1,1), PIntegerType.new(5,5), PIntegerType.new(9,9)]))
+  end
+
+  it 'can chain reverse_each after step on Enumerator' do
```
Enumerator?

hlindberg commented on spec/unit/pops/types/iterable_spec.rb:
```diff
@@ -0,0 +1,229 @@
+  it 'will produce the same result for each as for reverse_each.reverse_each' do
+    x1 = Iterable.on(5)
+    x2 = Iterable.on(5)
+    expect(x1.inject([]) { |a,i| a << i; a}).to eq(x2.reverse_each.reverse_each.inject([]) { |a,i| a << i; a})
```
always use `reduce` instead of `inject`

hlindberg commented on spec/unit/pops/types/type_calculator_spec.rb:
```diff
@@ -1965,20 +1981,20 @@ class Foo
   context 'when asking for an enumerable ' do
```
asking for an iterable

hlindberg commented on spec/unit/pops/types/type_calculator_spec.rb:
```diff
@@ -1965,20 +1981,20 @@ class Foo
   context 'when asking for an enumerable ' do
     it 'should produce an enumerable for an Integer range that is not infinite' do
```
enumerable

hlindberg commented:
I am fine with the general approach now!
I think that I made several comments where I am complaining about type inference being lossy, that may be because I don't understand the exact purpose of methods (I later found examples that showed that I got it wrong in some cases.

I will try to write some example in .pp that should illustrate more advanced type inference.

nfagerlund commented:
@hlindberg Hey, sorry for the delay. You asked about 4x functions vs. 4+3x for docs — yes, we still need a dummy 3x function for docs. This will have to continue until Puppet Strings can run cleanly against an unmodified checkout of the Puppet source; right now it results in a bit of a mess, so we're unable to rewrite our references tooling to rely on it.

hlindberg commented:
@nfagerlund are there tickets logged for what is missing? We want to fix those instead of doing the busy work of maintaining two versions (since it is not just a copy paste operation due to different formats)

hlindberg commented:
The documentation for each of the functions that support Iterable must be changed to reflect this.

hlindberg commented:
For documentation, I suggest we do not expand on the concept of being Iterable in each of the functions. Only what the rules are for that function. I suggest that we mention all the concrete things that can be iterated over, and say, something like "you can iterate over Array, Hash, ....., i.e. values that are of Iterable type". Later we can update with links to docs site that have the description of the "Iterable concept", and/or adjust the phrasing (Docs needs to be involved then). 

The functions must describe that "When the function is called without a block it will return a value of  Iterable[T] type that can be passed around as a first order data value, and later be iterated on by giving it as an argument to a function that accepts an Iterable[T]. Some returned Iterable[T] values will be the original LHS value, others will return an Iterable value that is of type Iterator[T]". Again, leave out the long explanation, what I put here in quotes should be enough (or something to that effect).

The documentation should also mention that the "snaking" of values is lazy as this is important for the order of evaluation; you may otherwise believe that a map followed by reduce will first call the map block n times, then the reduce block n times. I.e. "The block for each chained iteration function is called in succession for each value when it is consumed at the end of the chain".

The documentation has to show what happens when iterating over Hashes, What the rules are wrt tuples/hash.

I think there should be one example in each that shows iterable without block. Something simple like
$x = 3.each()
notice $x.map |$x| { $x * 10 }
# will notice [0, 10, 20]

Where the example uses the function in question on the first row

thallgren commented:
The functions cannot be called without a block though. This has not changed. Only new functions like `step` and `reverse_each` (in separate PR's) can do that and it's mentioned in their documentation.

In essence, nothing has changed in the behavior of the current iterative functions besides a more stringent check that the argument is an instance of the Iterable type rather than an instance of Any. They all return exactly the same result as before (none of them returns a transformed `Iterable`).

thallgren commented:
I would suggest that instead of repeating the long list of concrete `Iterable` objects in each function, we should say something short like "Array, Hash, or other Iterable object". It will be very cumbersome otherwise, to update each and every function when adding a new object that one can iterate over (and to remember that it must be done). It's also very confusing to mention that a String and an Integer are iterable without further explaining _how_ they are iterable. IMO, we must either leave them out and mention the Iterable as a concept explained elsewhere, or explain that concept it in every function.

hlindberg commented:
ok, clear on documentation - needs to be short, it snowballs otherwise.

nfagerlund commented:
@hlindberg https://tickets.puppetlabs.com/browse/PDOC-59 is the epic. I don't think I know about all the problems Strings runs into when you point it at core, but I captured several.

hlindberg commented:
Thanks @nfagerlund, not sure why my radar has not picked up those tickets. Will look into them. We need to fix those (or as you pointed out, fix the documentation comments/style of the code it is documenting if it does not follow the "standard way").

hlindberg commented:
Thanks @nfagerlund PDOC tickets were missing from our radar screen. We now see them and will try to do something about them asap.
